sdk/go/common/workspace/templates_zip.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workspace
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// Sanitize archive file pathing from "G305: Zip Slip vulnerability"
+func sanitizeArchivePath(d, t string) (v string, err error) {
+	v = filepath.Join(d, t)
+	if strings.HasPrefix(v, filepath.Clean(d)) {
+		return v, nil
+	}
+
+	return "", fmt.Errorf("%s: %s", "content filepath is tainted", t)
+}
+
+func isZIPTemplateURL(templateNamePathOrURL string) bool {
+	parsedURL, _ := url.Parse(templateNamePathOrURL)
+	return parsedURL.Path != "" && strings.HasSuffix(parsedURL.Path, ".zip")
+}
+
+func retrieveZIPTemplates(templateURL string) (TemplateRepository, error) {
+	var err error
+	// Create a temp dir.
+	var temp string
+	if temp, err = os.MkdirTemp("", "pulumi-template-"); err != nil {
+		return TemplateRepository{}, err
+	}
+
+	parsedURL, err := url.Parse(strings.ReplaceAll(templateURL, ".zip", ""))
+	if err != nil {
+		return TemplateRepository{}, err
+	}
+
+	var fullPath string
+	if fullPath, err = RetrieveZIPTemplateFolder(parsedURL, temp); err != nil {
+		return TemplateRepository{}, fmt.Errorf("failed to retrieve zip archive: %w", err)
+	}
+
+	return TemplateRepository{
+		Root:         temp,
+		SubDirectory: fullPath,
+		ShouldDelete: true,
+	}, nil
+}
+
+func RetrieveZIPTemplateFolder(templateURL *url.URL, tempDir string) (string, error) {
+	packageRequest, err := http.NewRequest(http.MethodGet, templateURL.String(), bytes.NewReader([]byte{}))
+	if err != nil {
+		return "", err
+	}
+	packageRequest.Header.Set("Accept", "application/zip")
+	requestQuery := packageRequest.URL.Query()
+	for key, values := range templateURL.Query() {
+		for _, value := range values {
+			requestQuery.Add(key, value)
+		}
+	}
+	packageRequest.URL.RawQuery = requestQuery.Encode()
```

Comment:
This looks like it's reading out and rewriting the query params, but if they're set, that means they're already set in `templateURL`. Does a test break if we remove this?

Reply:
It's reading the query params from the original request URL and setting them on the new one (without the .zip suffix).

Reply:
Ah - I see what you mean. I think that's correct, actually.  I don't have a test for url.Parse, so none will fail here, but I can test by removing this and requesting a non-default model from p-ai.

Reply:
Yep - confirmed this is a no-op:
<img width="1365" alt="image" src="https://github.com/pulumi/pulumi/assets/4128006/7d5f5649-ab0c-41fb-b8fc-10b0de42442e">